rotkehlchen/tests/unit/decoders/test_cowswap.py

```diff
@@ -0,0 +1,366 @@
+
+
+@pytest.mark.vcr()
+@pytest.mark.parametrize('ethereum_accounts', [[
+    '0x0D2f07876685bEcd81DDa1C897f2D6Cacc733fc1',
+    '0x31b6020CeF40b72D1e53562229c1F9200d00CC12',
+]])
+def test_2_swaps_token_to_token(database, ethereum_inquirer, ethereum_accounts):
+    tx_hex = deserialize_evm_tx_hash('0xd4d16ea74bbf806715f5f0e799fd5e8befbf369a9e5461fa9c0ed88d72bd06e4')  # noqa: E501
+    evmhash = deserialize_evm_tx_hash(tx_hex)
+    user_address_1, user_address_2 = ethereum_accounts
+    events, _ = get_decoded_events_of_transaction(
+        evm_inquirer=ethereum_inquirer,
+        database=database,
+        tx_hash=tx_hex,
+    )
+    expected_events = [
+        HistoryBaseEntry(
+            event_identifier=evmhash,
+            sequence_index=4,
+            timestamp=TimestampMS(1676976635000),
+            location=Location.ETHEREUM,
+            event_type=HistoryEventType.TRADE,
+            event_subtype=HistoryEventSubType.SPEND,
+            asset=A_WBTC,
+            balance=Balance(amount=FVal('0.15463537')),
+            location_label=user_address_2,
+            notes='Swap 0.15463537 WBTC in cowswap',
+            counterparty=CPT_COWSWAP,
+        ), HistoryBaseEntry(
+            event_identifier=evmhash,
+            sequence_index=5,
+            timestamp=TimestampMS(1676976635000),
+            location=Location.ETHEREUM,
+            event_type=HistoryEventType.TRADE,
+            event_subtype=HistoryEventSubType.RECEIVE,
+            asset=Asset('eip155:1/erc20:0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48'),
+            balance=Balance(amount=FVal('3800')),
+            location_label=user_address_2,
+            notes='Receive 3800 USDC as the result of a swap in cowswap',
+            counterparty=CPT_COWSWAP,
+        ), HistoryBaseEntry(
+            event_identifier=evmhash,
+            sequence_index=7,
+            timestamp=TimestampMS(1676976635000),
+            location=Location.ETHEREUM,
+            event_type=HistoryEventType.TRADE,
+            event_subtype=HistoryEventSubType.SPEND,
+            asset=Asset('eip155:1/erc20:0xe9B076B476D8865cDF79D1Cf7DF420EE397a7f75'),
+            balance=Balance(amount=FVal('16000')),
+            location_label=user_address_1,
+            notes='Swap 16000 FUND in cowswap',
+            counterparty=CPT_COWSWAP,
+        ), HistoryBaseEntry(
+            event_identifier=evmhash,
+            sequence_index=8,
+            timestamp=TimestampMS(1676976635000),
+            location=Location.ETHEREUM,
+            event_type=HistoryEventType.TRADE,
+            event_subtype=HistoryEventSubType.RECEIVE,
+            asset=A_WETH,
+            balance=Balance(amount=FVal('4.870994011222719015')),
+            location_label=user_address_1,
+            notes='Receive 4.870994011222719015 WETH as the result of a swap in cowswap',  # noqa: E501
+            counterparty=CPT_COWSWAP,
+        ), HistoryBaseEntry(
```

Comment:
What is this approval at the end? It's for both tests. Why? What is the `0xC92E8bdf79f0507f65a392b0ab4667716BFE0110` contract and why is this at the end?

Reply:
1. `0xC92E8bdf79f0507f65a392b0ab4667716BFE0110` is `GPv2VaultRelayer`, which is used by cowswap to spend users' tokens. Example: https://ethtx.info/0xd4d16ea74bbf806715f5f0e799fd5e8befbf369a9e5461fa9c0ed88d72bd06e4/ 
![image](https://user-images.githubusercontent.com/63492346/223659753-695f0874-f400-417e-9ad7-c36ac5a9cd7f.png)

3. The approval event may be or may not be emitted depending on ERC20 implementation of tokens. For example, `FUND` token https://etherscan.io/token/0xe9b076b476d8865cdf79d1cf7df420ee397a7f75#code does emit an approval event on each call to `transferFrom`.
```solidity
function transferFrom(
        address sender,
        address recipient,
        uint256 amount
    ) public virtual override returns (bool) {
        _transfer(sender, recipient, amount);

        uint256 currentAllowance = _allowances[sender][_msgSender()];
        require(currentAllowance >= amount, "ERC20: transfer amount exceeds allowance");
        unchecked {
            _approve(sender, _msgSender(), currentAllowance - amount);
        }

        return true;
    }
```

Reply:
I see. So that's non-standard behaviour on the side of the token used in this test. Though it's an underlying library so may happen to more. Oh god so silly.

If you check it's not really approvign more. It's approving less. Which is correct behaviour.

The way approvals work is you approve an address (in this case cowswap) to spend your tokens. This is how all AMMs and other contracts work. You approve a certain amount. When it transfers some of it, the contract has to reduce the approval by the spent amount. But it should not / does not need to emit an event when this happens.

Seems they do here and we catch it and show it. Which is silly as for us it's just noise.

Perhaps we should update the text of approvals to indicate that the "approved amount of token X to be spent by address Y is now set to Z".